File: DSA With C/Using Linked List/circular_linked_list.c
//circular linked list
#include<stdio.h>
#include<stdlib.h>
struct node
{
    int data;
    struct node *next;       
}*head;
int c;
void create(int);
void display();
void deleteatbeg();
void deleteatend();
void deleteatany(int);
void insertatbeg();
void insertatend();
void insertatany();
void search();
void main()
{
    int n,ch,data;
    while(1)
    {
        printf("\n*MENU*\n1. Create list\n2. Delete at begining\n3. Delete at end\n4. Delete at any\n5. insert at begining\n6. insert at end\n7. insert at any\n8. search\n9. Exit\n\nEnter your choice : ");
        scanf("%d",&ch);//1 2
        switch(ch)
        {
            case 1 :
                printf("Enter the number of nodes : ");
                scanf("%d",&n);//n=3
                create(n);//calling 
                break;
            case 2 :
                deleteatbeg();
                break;
            case 3 :
                deleteatend();
                break; 
            case 4 :                    
                deleteatany(c);
                break;
            case 5 :
                insertatbeg();
                break;
            case 6 :
                insertatend();
                break; 
            case 7 : 
                insertatany();   //5                 
                break; 
            case 8 : 
                search();                     
                break;      
            case 9 :
                exit(0);
            default :
                printf("\nWrong choice!!!\n");                     
        }
    }
}
void create(int x)//n=3
{
    struct node *n,*temp;
    for(int i=1;i<=x;i++)//i=2;4<=3;i++
    {
        n=(struct node*)malloc(sizeof(struct node));//100-->132
        if(n == NULL)
        {
            printf("\nMemory can't be allocated\n");
            return;
        }
        if(i == 1)
            temp = head = n;
        printf("Enter the data : ");
        scanf("%d",&n->data);     
        n->next = head;           
        temp->next = n;           
        temp = n;                               
    }
    display();
}
void display()
{
    c=0;
    struct node *temp1=head;//*temp1=100
    if(head==NULL)//100==NULL
    { 
        printf("List is empty\n");
        return;
    }
    printf("\nData in the list : \n\n");           
    do
    {
        c++;//c=3                
        printf("Data = %d and Address = %u\n",temp1->data,temp1);
        temp1=temp1->next;//temp1=100                 
    }
    while(temp1!=head);//circular  100!=100  
    printf("\nThe number of nodes are : %d\n",c);       //3
}
void deleteatbeg()
{
    struct node *end=head;
    if(end->next==end)
    {
        head=NULL;
        free(end);
    }
    else
    {
        while(end->next!=head)
            end = end->next;
        head = head->next;
        free(end->next);
        end->next = head;     
    }
    printf("\nDeletion Successful...\n");
    display();
}
void deleteatend()
{
    struct node *end = head;
    if(end->next == end)
    {
        head = NULL;
        free(end);
    }
    else
    {
        while((end->next)->next!=head)       
            end=end->next;
        free(end->next);
        end->next = head; 
    }   
    printf("\nDeletion Successful...\n");
    display();
}
void deleteatany(int n) 
{
    struct node *end ;
    end = head;
    int p, i;
    printf("\nEnter the position : ");
    scanf("%d", &p);
    if (p < 1 || p > n) 
    {
        printf("Invalid position!!\n");
        return;
    }
    else if (p == 1) // Delete at beginning
        deleteatbeg();
    else if (p == n) // Delete at end
        deleteatend();
    struct node *prev = NULL;
    for (i = 1; i < p; i++,end = end->next) 
        prev = end;
    prev->next = end->next;
    free(end);
    printf("\nDeletion Successful...\n");
    display();
}
void insertatbeg()
{
    struct node *n,*end=head;//end=100    
    do
        end=end->next;//end=132                 
    while(end->next!=head);//100!=100   
    n = (struct node*)malloc(sizeof(struct node)); 
    printf("Enter the data : ");
    scanf("%d",&n->data);//x=40
    n->next=head;
    head = end->next=n;//132->next=head=148
    display();
}
void insertatend()
{
    struct node *n,*end=head;//end=148
    do
        end=end->next;//end=132          
    while(end->next!=head);//148!=148   
    n = (struct node*)malloc(sizeof(struct node)); 
    printf("Enter the data : ");
    scanf("%d",&n->data);
    n->next=head;
    end->next=n;//132->next= 164
    display();
}
void insertatany()//s=5
{
    struct node *n3,*t=head;
    int i,p;
    printf("Enter the position :");
    scanf("%d",&p);//n=3
    if(p==1)
        insertatbeg();
    else if(p==c+1)
        insertatend();
    else if(p>c+1) 
        return;
    for(i=2;i<p;i++)//i=1;1<=3;i++              
        t=t->next;//t=100                 
    n3=(struct node*)malloc(sizeof(struct node));
    printf("Enter your data : ");
    scanf("%d",&n3->data);
    n3->next=t->next;
    t->next=n3;
    display();  
}
void search()
{
    int x,f=0;
    struct node *temp2 = head ;
    printf("\nEnter the no. to search : ");
    scanf("%d",&x);
    if(head==NULL)
    {
        printf("Unable to find!!! No node in the linked list.");
        return;
    }
    do
    {
        if(temp2->data == x)
        {
            f=1;
            break;
        }
        temp2=temp2->next;
    }
    while(temp2!=head);		
    (f!=1)?printf("\nSearch not found !!\n"):printf("\nSearch found..\n");
}

File: DSA With Java/Code/using linked list/CircularLinkedList.java
import java.util.Scanner;
class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}
public class CircularLinkedList {
    private Node head = null;
    private Node tail = null;

    public void createList() {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of nodes: ");
        int n = sc.nextInt();

        for (int i = 1; i <= n; i++) {
            System.out.print("Enter data for node " + i + ": ");
            int data = sc.nextInt();
            insertAtEnd(data);
        }
        System.out.println("Circular linked list created successfully!");
        sc.close();
    }
    public void traverseList() {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }
        Node current = head;
        System.out.print("Circular Linked List: ");
     do {
            System.out.print(current.data + " ");
            current = current.next;
        } while (current != head);
        System.out.println();
    }
    public void insertAtBeginning(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            tail.next = head;
        } else {
            newNode.next = head;
            head = newNode;
            tail.next = head;
        }
    }
    public void insertAtEnd(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            tail.next = head;
        } else {
            tail.next = newNode;
            tail = newNode;
            tail.next = head;
        }
    }
    public void insertAtSpecific(int pos, int data) {
        if (pos <= 0) {
            System.out.println("Invalid position!");
            return;
        }

        Node newNode = new Node(data);

        if (pos == 1) {
            insertAtBeginning(data);
            return;
        }

        Node temp = head;
        int count = 1;
        while (count < pos - 1 && temp.next != head) {
            temp = temp.next;
            count++;
        }
        newNode.next = temp.next;
        temp.next = newNode;
        if (temp == tail) {
            tail = newNode;
        }
    }
    public void deleteAtBeginning() {
        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return;
        }

        if (head == tail) {
            head = tail = null;
        } else {
            head = head.next;
            tail.next = head;
        }
        System.out.println("Node deleted from beginning.");
    }
    public void deleteAtEnd() {
        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return;
        }

        if (head == tail) {
            head = tail = null;
        } else {
            Node temp = head;
            while (temp.next != tail) {
                temp = temp.next;
            }
            temp.next = head;
            tail = temp;
        }
        System.out.println("Node deleted from end.");
    }
    public void deleteAtSpecific(int pos) {
        if (head == null) {
            System.out.println("List is empty. Nothing to delete.");
            return;
        }

        if (pos <= 0) {
            System.out.println("Invalid position!");
            return;
        }

        if (pos == 1) {
            deleteAtBeginning();
            return;
        }

        Node temp = head;
        Node prev = null;
        int count = 1;

        while (count < pos && temp.next != head) {
            prev = temp;
            temp = temp.next;
            count++;
        }

        if (count != pos) {
            System.out.println("Position out of range!");
            return;
        }

        prev.next = temp.next;

        if (temp == tail) {
            tail = prev;
        }

        System.out.println("Node deleted from position " + pos + ".");
    }
    public void search(int key) {
        if (head == null) {
            System.out.println("List is empty.");
            return;
        }

        Node temp = head;
        int pos = 1;
        do {
            if (temp.data == key) {
                System.out.println("Element " + key + " found at position " + pos + ".");
                return;
            }
            temp = temp.next;
            pos++;
        } while (temp != head);

        System.out.println("Element " + key + " not found in the list.");
    }
    public static void main(String[] args) {
        CircularLinkedList list = new CircularLinkedList();
        Scanner sc = new Scanner(System.in);
        int choice, data, pos;

        do {
            System.out.println("\n====== Circular Linked List Menu ======");
            System.out.println("\t1. Create List");
            System.out.println("\t2. Traverse List");
            System.out.println("\t3. Insert at Beginning");
            System.out.println("\t4. Insert at End");
            System.out.println("\t5. Insert at Specific Position");
            System.out.println("\t6. Delete at Beginning");
            System.out.println("\t7. Delete at End");
            System.out.println("\t8. Delete at Specific Position");
            System.out.println("\t9. Search an Element");
            System.out.println("\t10. Exit");
            System.out.print("\tEnter your choice: ");
            choice = sc.nextInt();

            switch (choice) {
                case 1:
                    list.createList();
                    break;
                case 2:
                    list.traverseList();
                    break;
                case 3:
                    System.out.print("Enter data: ");
                    data = sc.nextInt();
                    list.insertAtBeginning(data);
                    break;
                case 4:
                    System.out.print("Enter data: ");
                    data = sc.nextInt();
                    list.insertAtEnd(data);
                    break;
                case 5:
                    System.out.print("Enter position: ");
                    pos = sc.nextInt();
                    System.out.print("Enter data: ");
                    data = sc.nextInt();
                    list.insertAtSpecific(pos, data);
                    break;
                case 6:
                    list.deleteAtBeginning();
                    break;
                case 7:
                    list.deleteAtEnd();
                    break;
                case 8:
                    System.out.print("Enter position: ");
                    pos = sc.nextInt();
                    list.deleteAtSpecific(pos);
                    break;
                case 9:
                    System.out.print("Enter element to search: ");
                    data = sc.nextInt();
                    list.search(data);
                    break;
                case 10:
                    System.out.println("Exiting program...");
                    break;
                default:
                    System.out.println("Invalid choice! Please try again.");
            }
        } while (choice != 10);
        sc.close();
    }
}